Allow multiple summaries per pull request

lagru opened the pull request:
Tackles the first case in https://github.com/scientific-python/changelist/issues/45:

> **1pr-to-many**: A pull requests addresses multiple things which should ideally appear as multiple summaries. Possibly even in separate sections.

It would be good to have this in before the next release of scikit-image. So that changelist can gracefully deal with https://github.com/scikit-image/scikit-image/pull/6695.

lagru commented on README.md:
````diff
@@ -93,7 +95,7 @@ ignored_user_logins = [
 # If you modify this regex, make sure to match the content with a capture
 # group named "summary".
-pr_summary_regex = "^```release-note\\s*(?P<summary>[\\s\\S]*?\\w[\\s\\S]*?)\\s*^```"
+pr_summary_regex = "^```release-note\\s*((?P<label>[^:]*):)?(?P<summary>[\\s\\S]*?\\w[\\s\\S]*?)\\s*^```"
````
As a start I'm proposing the following syntax (ignore escaping `\`):

```
\```release-note
Enhancement: Add new `mask` parameter to `foo` to allow masking of values.
\```

\```release-note
Bug fix: Fix endless loop in `foo` if NaN is passed as input. 
\```
```
These two will be translated into two items, one in the "Enhancements" and one in the "Bug fixes" section. If a preceding `Label: ` isn't given, the labels of the PR are used.

stefanv replied:
What I was trying to say is that, unfortunately, contributors cannot see the fenced code block without editing the PR description. But, with the release notes heading you added, that's not such a big problem.

I still think this might be a bit brittle, since colons are so commonly used in descriptions of all sorts. However, if it's limited to "SINGLE-WORD: stuff" it seems less likely. So, let's give it a shot.

lagru replied:
Hmm, it's currently not limited to single words infront of the colon. What do you think about a compromise that uses two colons?

```
\```release-note
Enhancement:: Add new `mask` parameter to `foo` to allow masking of values.
\```

\```release-note
Bug fix :: Fix endless loop in `foo` if NaN is passed as input. 
\```
```
Alternatively we could just move the `{label="release-note"}` into the next line so that it's visible. E.g. 
```
{label="release-note"} Some note.
```
What I like about this approach is that the `label=` part makes it more expressive and also that the position doesn't matter.

lagru commented on src/changelist/_format.py:
```diff
@@ -3,21 +3,71 @@
+            matches = tuple(pr_summary_regex.finditer(pr.body))
+            if not matches:
+                logger.debug("falling back to title for %s", pr.html_url)
+                matches = [{"summary": pr.title, "label": None}]
```
If `LABEL: ` is given in the PR title (e.g. the convention in the NumPy) and changelist falls back to it for the summary, the behavior is unchanged and it's included directly.

lagru commented on src/changelist/_format.py:
```diff
@@ -3,21 +3,71 @@
 logger = logging.getLogger(__name__)
 
 
+@dataclass(frozen=True, slots=True)
+class ChangeItem:
+    content: str
+    reference_name: str
+    reference_url: str
+    labels: tuple[str]
+    timestamp: datetime
```
This new data class (`PullRequest` -> `ChangeItem`) uncouples `_format.py` and `_query.py` a bit so that `MdFormatter` doesn't have to know about PRs at all. 

I intend to do the same for `NamedUser` -> `UserItem`. This should make it way easier to test the formatter without mocking. Because the PyGitHub objects are kinda opaque and not easy to initialize.

Long-term, might also help if the query part is moved to GraphQL queries.

lagru reviewed (commented):
FYI, this is a first working draft and still rough around the edges / missing documentation. But see the comments for some early discussion & context.

lagru commented:
Okay, finally Python 3.9 is happy (sorry for the notification noise). I think this is ready to merge.

As a real world example, this
```
changelist scikit-image/scikit-image 54316d5~1 54316d5
```
includes just the merge commit from https://github.com/scikit-image/scikit-image/pull/6695 and creates the following notes:

<details><summary>Details</summary>

```md
# scikit-image x.y.z

We're happy to announce the release of scikit-image x.y.z!

## New Features

- Add parameter `mode` to `binary_erosion`, `binary_dilation`, `binary_opening` and `binary_closing` in`skimage.morphology`. These new parameters determine how array borders are handled ([#6695](https://github.com/scikit-image/scikit-image/pull/6695)).
- Add parameters `mode` and `cval` to `erosion`, `dilation`, `opening`, `closing`, `white_tophat`, and `black_tophat` in `skimage.morphology`. These new parameters determine how array borders are handled ([#6695](https://github.com/scikit-image/scikit-image/pull/6695)).
- Add functions `mirror_footprint` and `pad_footprint` to `skimage.morphology` ([#6695](https://github.com/scikit-image/scikit-image/pull/6695)).

## API Changes

- Parameters `shift_x` and `shift_y` in `skimage.morphology.erosion` and `skimage.morphology.dilation` are deprecated. Use `pad_footprint` or modify the footprint manually instead ([#6695](https://github.com/scikit-image/scikit-image/pull/6695)).

## Bug Fixes

- Ensure `skimage.morphology.closing` and `skimage.morphology.opening` are extensive and anti-extensive, respectively, if the footprint is not mirror symmetric ([#6695](https://github.com/scikit-image/scikit-image/pull/6695)).

## Contributors

2 authors added to this release (alphabetically):

- Cris Luengo ([@crisluengo](https://github.com/crisluengo))
- Lars Grüter ([@lagru](https://github.com/lagru))

3 reviewers added to this release (alphabetically):

- Cris Luengo ([@crisluengo](https://github.com/crisluengo))
- Lars Grüter ([@lagru](https://github.com/lagru))
- Stefan van der Walt ([@stefanv](https://github.com/stefanv))

_These lists are automatically generated, and may not be complete or may contain
duplicates._
```

</details> 

Let me know what you think. :)

lagru commented:
@stefanv, could I ask you to have a look at this if you have the time? :pray: Not the highest priority in my opinion but I'd hate to rush this just before our next skimage release.

stefanv reviewed (commented):
Hi @lagru, this looks good overall. I suggested a syntax change to how the label is indicated.

lagru commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,33 @@
+        with:
+          path: ~/.cache/pip
+          key: ${{ runner.os }}-pip-${{ hashFiles('**/pyproject.toml') }}
+          restore-keys: |
+            ${{ runner.os }}-pip-
```
I just stole the [workflow from spin](https://github.com/scientific-python/spin/blob/f43a217af4a39d0383738c60197f2b802d6871aa/.github/workflows/lint.yml#L24) and modified it to fit. :sweat_smile: 

Looking at [`actions/cache`'s documentation](https://github.com/actions/cache#inputs) it looks like it might be useful to keep it around.

stefanv replied:
*those* people :roll_eyes: :joy:

stefanv commented on README.md:
````diff
@@ -83,17 +83,21 @@ ignored_user_logins = [
 # ```release-note
-# An ideally expressive description of the change that is included as a single
-# bullet point. Newlines are removed.
+# Label: An ideally expressive description of the change that is included as
````
*those* people :roll_eyes: :joy:

lagru replied:
Re https://github.com/scientific-python/changelist/pull/53#issuecomment-1832391609, still need to address this.

lagru commented on pyproject.toml:
```diff
@@ -5,6 +5,7 @@ build-backend = "setuptools.build_meta"
 [project]
 name = "changelist"
 version = "0.5rc0.dev0"
+# TODO add slots=True, kw_only=True to dataclasses starting with >=3.10
```
Support for `slots` and `kw_only` in `dataclass` was only added in Python 3.10. Because changelist still supports 3.9 we can't add them yet.

However, I'd like to add have them eventually so I added a reminder. IMO, `kw_only` forces more expressive calls to init, and `slots` safes a bit of memory at no cost.

stefanv commented:
@lagru Let me know if you intend to make any other changes, otherwise I'll merge.

lagru commented:
@stefanv. I've addressed all your comments and switched to the `{label="Some label"}` based syntax. This `{...}` block can be added anywhere once in the summary. So it can be used as a [fenced_code_attributes](https://pandoc.org/MANUAL.html#extension-fenced_code_attributes) that is hidden, or anywhere else so that it's visible. 

I've also added basic tests for this in `test_objects.py`.

lagru commented:
As we talked about, I've added a section in the README to document the new feature and make it clearer how to use it. It's duplicating the documentation in the config a little bit, but it's probably less hidden this way and should read nicer (no talk of regex).

stefanv reviewed (approved):
Looks great! Is overriding the title documented?

lagru commented on README.md:
```diff
@@ -33,9 +35,36 @@ changelist scientific-python/changelist v0.2.0 main
 
+## Writing pull request summaries
+
+By default, changelist will fall back to the title of a pull request and its
+GitHub labels to sort it into the appropriate section. However, if you want
+longer summaries of your changes you can add a code block with the following
+form anywhere in the description of the pull request:
```
@stefanv, re https://github.com/scientific-python/changelist/pull/53#pullrequestreview-1846584015: yes, overriding the title is documented.

stefanv commented on README.md:
````diff
@@ -33,9 +35,36 @@ changelist scientific-python/changelist v0.2.0 main
+    a single bullet point. Newlines are removed.
+    ```
+
+Sometimes pull requests introduce changes that should be listed in different
````
```suggestion
Sometimes pull requests introduce multiple changes that should be listed in different
```

stefanv commented:
Thanks Lars!
